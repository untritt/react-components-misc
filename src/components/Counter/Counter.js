import _ from "lodash";
import React from "react";

export default class Counter extends React.Component {
  state = {
    log: []
  };

  getFnHandler = fn => e => {
    e.preventDefault();

    this.setState(({ log }) => {
      const newValue = log[0] ? fn(log[0].value) : fn(0);
      const newItem = {
        value: newValue,
        key: _.uniqueId()
      };
      return {
        log: [newItem, ...log]
      };
    });
  };

  onInc = this.getFnHandler(x => x + 1);
  onDec = this.getFnHandler(x => x - 1);

  getClickHandler = key => e => {
    e.preventDefault();
    this.setState(({ log }) => ({
      log: log.filter(item => item.key !== key)
    }));
  };

  renderLog() {
    if (this.state.log.length === 0) {
      return null;
    }
    return (
      <ul className="list-group">
        {this.state.log.map(({ key, value }) => (
          <li
            key={key}
            className="list-group-item"
            onClick={this.getClickHandler(key)}
          >
            <a href="#">{value}</a>
          </li>
        ))}
      </ul>
    );
  }

  render() {
    return (
      <div>
        <div className="btn-group" role="group">
          <button className="btn btn-secondary inc" onClick={this.onInc}>
            +
          </button>
          <button className="btn btn-secondary dec" onClick={this.onDec}>
            -
          </button>
        </div>
        {this.renderLog()}
      </div>
    );
  }
}
