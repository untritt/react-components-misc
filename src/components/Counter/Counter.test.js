import React from "react";
import Enzyme, { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Counter from "../Counter";

Enzyme.configure({ adapter: new Adapter() });

test("Counter", () => {
  const wrapper = mount(<Counter />);
  expect(wrapper.render()).toMatchSnapshot();

  const inc = wrapper.find(".inc");
  const dec = wrapper.find(".dec");

  inc.simulate("click");
  expect(wrapper.render()).toMatchSnapshot();

  inc.simulate("click");
  expect(wrapper.render()).toMatchSnapshot();

  dec.simulate("click");
  expect(wrapper.render()).toMatchSnapshot();

  inc.simulate("click");
  expect(wrapper.render()).toMatchSnapshot();

  dec.simulate("click");
  expect(wrapper.render()).toMatchSnapshot();

  const items = wrapper.find("li.list-group-item a");
  items.at(3).simulate("click");
  expect(wrapper.render()).toMatchSnapshot();

  items.at(1).simulate("click");
  expect(wrapper.render()).toMatchSnapshot();

  items.at(2).simulate("click");
  expect(wrapper.render()).toMatchSnapshot();

  inc.simulate("click");
  expect(wrapper.render()).toMatchSnapshot();

  dec.simulate("click");
  expect(wrapper.render()).toMatchSnapshot();
});
