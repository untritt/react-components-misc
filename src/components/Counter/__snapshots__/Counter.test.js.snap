// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter 1`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary inc"
    >
      +
    </button>
    <button
      class="btn btn-secondary dec"
    >
      -
    </button>
  </div>
</div>
`;

exports[`Counter 2`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary inc"
    >
      +
    </button>
    <button
      class="btn btn-secondary dec"
    >
      -
    </button>
  </div>
  <ul
    class="list-group"
  >
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
  </ul>
</div>
`;

exports[`Counter 3`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary inc"
    >
      +
    </button>
    <button
      class="btn btn-secondary dec"
    >
      -
    </button>
  </div>
  <ul
    class="list-group"
  >
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        2
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
  </ul>
</div>
`;

exports[`Counter 4`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary inc"
    >
      +
    </button>
    <button
      class="btn btn-secondary dec"
    >
      -
    </button>
  </div>
  <ul
    class="list-group"
  >
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        2
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
  </ul>
</div>
`;

exports[`Counter 5`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary inc"
    >
      +
    </button>
    <button
      class="btn btn-secondary dec"
    >
      -
    </button>
  </div>
  <ul
    class="list-group"
  >
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        2
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        2
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
  </ul>
</div>
`;

exports[`Counter 6`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary inc"
    >
      +
    </button>
    <button
      class="btn btn-secondary dec"
    >
      -
    </button>
  </div>
  <ul
    class="list-group"
  >
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        2
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        2
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
  </ul>
</div>
`;

exports[`Counter 7`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary inc"
    >
      +
    </button>
    <button
      class="btn btn-secondary dec"
    >
      -
    </button>
  </div>
  <ul
    class="list-group"
  >
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        2
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
  </ul>
</div>
`;

exports[`Counter 8`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary inc"
    >
      +
    </button>
    <button
      class="btn btn-secondary dec"
    >
      -
    </button>
  </div>
  <ul
    class="list-group"
  >
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
  </ul>
</div>
`;

exports[`Counter 9`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary inc"
    >
      +
    </button>
    <button
      class="btn btn-secondary dec"
    >
      -
    </button>
  </div>
  <ul
    class="list-group"
  >
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
  </ul>
</div>
`;

exports[`Counter 10`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary inc"
    >
      +
    </button>
    <button
      class="btn btn-secondary dec"
    >
      -
    </button>
  </div>
  <ul
    class="list-group"
  >
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        2
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
  </ul>
</div>
`;

exports[`Counter 11`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn btn-secondary inc"
    >
      +
    </button>
    <button
      class="btn btn-secondary dec"
    >
      -
    </button>
  </div>
  <ul
    class="list-group"
  >
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        2
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
    <li
      class="list-group-item"
    >
      <a
        href="#"
      >
        1
      </a>
    </li>
  </ul>
</div>
`;
